src/system.hpp

```diff
@@ -76,17 +97,102 @@ const size_t READ_BUFFER_SIZE = 65536;
 // _WIN32_WINNT is set in the generated header config.h
 #    error _WIN32_WINNT is undefined
 #  endif
+
+#  ifdef _MSC_VER
+typedef int mode_t;
+typedef int pid_t;
+#  endif
+
+#  ifndef __MINGW32__
+typedef int64_t ssize_t;
+#  endif
+
+// Defined in Win32Util.cpp
+void usleep(int64_t usec);
+struct tm* localtime_r(time_t* _clock, struct tm* _result);
+
+#  ifdef _MSC_VER
+int gettimeofday(struct timeval* tp, struct timezone* tzp);
+int asprintf(char** strp, const char* fmt, ...);
+#  endif
+
+// From:
+// http://mesos.apache.org/api/latest/c++/3rdparty_2stout_2include_2stout_2windows_8hpp_source.html
+#  ifdef _MSC_VER
+const mode_t S_IRUSR = mode_t(_S_IREAD);
+const mode_t S_IWUSR = mode_t(_S_IWRITE);
+#  endif
+
+// From https://stackoverflow.com/a/62371749/262458
+#  define _CRT_INTERNAL_NONSTDC_NAMES 1
+#  include <sys/stat.h>
+#  if !defined(S_ISREG) && defined(S_IFMT) && defined(S_IFREG)
+#    define S_ISREG(m) (((m)&S_IFMT) == S_IFREG)
+#  endif
+#  if !defined(S_ISDIR) && defined(S_IFMT) && defined(S_IFDIR)
+#    define S_ISDIR(m) (((m)&S_IFMT) == S_IFDIR)
+#  endif
+
+#  include <direct.h>
+#  include <io.h>
+#  include <process.h>
+#  define NOMINMAX 1
 #  include <windows.h>
-#  define mkdir(a, b) mkdir(a)
+#  define mkdir(a, b) _mkdir(a)
 #  define link(src, dst) (CreateHardLink(dst, src, nullptr) ? 0 : -1)
 #  define execv(a, b) win32execute(a, b, 0, -1, -1)
+#  define strncasecmp _strnicmp
+#  define strcasecmp _stricmp
+
+// Visual Studio has deprecated the non-underscore-prefixed names of some POSIX
+// functions.
+#  ifndef __GNUC__
+#    define open _open
```

Comment:
This is actually what triggered my previous comment about not being so sure. Is there no better way?

Redefining common symbols like `open`, `close`, `read` and `write` seems very ugly and error prone. For instance, say that we start using a third party library with a `C::open` method. Then it won't work to call `C::open` since the header will declare an `_open` method (which won't be defined in the library).

Reply:
I noticed that while this is true for GCC, the MSVC preprocessor is smart enough to not do that.

Regardless you're right that this is not particularly great, and I will remove this and just suppress this warning instead.

Since the non-underscore versions of these functions are deprecated, there may be a chance they will be dropped entirely by Microsoft at some point, however I don't think that will happen any time soon, like in the next decade or two.

Reply:
This is fixed.